astmio/modern_records.py

```diff
@@ -1,1254 +1,223 @@
 
-log = logging.getLogger(__name__)
+from .exceptions import InvalidFieldError
+from .logging import get_logger
 
+if TYPE_CHECKING:
+    from .models import RecordConfig
 
-class ASTMBaseRecord(BaseModel):
+log = get_logger(__name__)
+
+
+@dataclass(slots=True)
+class RecordMetadata:
     """
-    Enhanced base class for all ASTM records with comprehensive functionality.
+    A container for metadata attached to a dynamically generated record model.
 
-    Features:
-    - Robust validation and error handling
-    - Multiple serialization formats (JSON, XML, CSV, ASTM)
-    - Metadata tracking and audit trails
-    - Type-safe field access
-    - Performance optimizations
+    This metadata provides the bridge between the ASTM standard's positional
+    field structure and the named fields of the Pydantic model at runtime.
+    """
+
+    source_config: Union["RecordConfig", "ComponentField"]
+
+    # Bidirectional mapping for easy lookups
+    position_to_name: Dict[int, str] = field(default_factory=dict)
+    name_to_position: Dict[str, int] = field(default_factory=dict)
+
+    # Simple lists and dictionaries for runtime checks
+    required_fields: List[str] = field(default_factory=list)
+    field_types: Dict[str, str] = field(default_factory=dict)
+
+    def get_position(self, name: str) -> Optional[int]:
+        """Looks up the 1-based ASTM position for a given field name."""
+        return self.name_to_position.get(name)
+
+    def get_type(self, name: str) -> Optional[str]:
+        """Looks up the configured type string for a given field name."""
+        return self.field_types.get(name)
+
+
+class ASTMBaseRecord(BaseModel):
+    """
+    The final, definitive base class for all dynamically created Pydantic models.
+    It provides a robust parser for pre-processed data, audit metadata, and
+    essential serialization methods.
     """
 
     model_config = ConfigDict(
         str_strip_whitespace=True,
         validate_assignment=True,
         extra="forbid",
-        frozen=False,  # Allow modification for practical use
+        frozen=False,
         arbitrary_types_allowed=True,
-        use_enum_values=True,
     )
 
-    # Metadata fields (private attributes)
-    _created_at: Optional[datetime] = PrivateAttr(default_factory=datetime.now)
+    _created_at: datetime = PrivateAttr(default_factory=datetime.now)
     _updated_at: Optional[datetime] = PrivateAttr(default=None)
-    _validation_errors: List[str] = PrivateAttr(default_factory=list)
     _source: Optional[str] = PrivateAttr(default=None)
 
-    # Class-level configuration
-    _astm_field_mapping: ClassVar[Dict[str, int]] = {}
-    _required_fields: ClassVar[List[str]] = []
-    _max_field_lengths: ClassVar[Dict[str, int]] = {}
+    _astm_metadata: ClassVar["RecordMetadata"]
 
     def __setattr__(self, name: str, value: Any) -> None:
-        """Track updates for audit trail."""
+        """Override to automatically track updates for an audit trail."""
```

Comment:
This should be a plugin
```suggestion
        """Override to automatically track updates for an audit trail."""
```